query-engine/core/src/query_graph_builder/write/delete.rs

```diff
@@ -20,37 +21,71 @@ pub(crate) fn delete_record(
     let filter = extract_unique_filter(where_arg.value.try_into()?, &model)?;
 
-    // Prefetch read query for the delete
-    let mut read_query = read::find_unique(field, model.clone(), query_schema)?;
-    read_query.add_filter(filter.clone());
-
-    let read_node = graph.create_node(Query::Read(read_query));
-    let delete_query = Query::Write(WriteQuery::DeleteRecord(DeleteRecord {
-        model: model.clone(),
-        record_filter: Some(filter.into()),
-    }));
-
-    let delete_node = graph.create_node(delete_query);
-    utils::insert_emulated_on_delete(graph, query_schema, &model, &read_node, &delete_node)?;
-
-    graph.create_edge(
-        &read_node,
-        &delete_node,
-        QueryGraphDependency::ProjectedDataDependency(
-            model.primary_identifier(),
-            Box::new(|delete_node, parent_ids| {
-                if !parent_ids.is_empty() {
-                    Ok(delete_node)
-                } else {
-                    Err(QueryGraphBuilderError::RecordNotFound(
-                        "Record to delete does not exist.".to_owned(),
-                    ))
-                }
+    if can_use_atomic_delete(query_schema, &field) {
+        // Database supports returning the deleted row, so just the delete node will suffice.
+        let nested_fields = field.nested_fields.unwrap().fields;
+        let selected_fields = read::utils::collect_selected_scalars(&nested_fields, &model);
+        let selection_order = read::utils::collect_selection_order(&nested_fields);
+        let delete_query = Query::Write(WriteQuery::DeleteRecord(DeleteRecord {
+            name: field.name,
+            model,
+            record_filter: Some(filter.into()),
+            selected_fields: Some(DeleteRecordFields {
+                fields: selected_fields,
+                order: selection_order,
             }),
-        ),
-    )?;
+        }));
+        let delete_node = graph.create_node(delete_query);
+
+        // TODO laplab: figure out how do we update relations here.
+        graph.add_result_node(&delete_node);
+    } else {
+        // In case database does not support returning the deleted row, we need to emulate that
+        // behaviour by first reading the row and only then deleting it.
+        let mut read_query = read::find_unique(field, model.clone(), query_schema)?;
+        read_query.add_filter(filter.clone());
+        let read_node = graph.create_node(Query::Read(read_query));
+
+        let delete_query = Query::Write(WriteQuery::DeleteRecord(DeleteRecord {
+            name: String::new(),
+            model: model.clone(),
+            record_filter: Some(filter.into()),
+            selected_fields: None,
+        }));
+        let delete_node = graph.create_node(delete_query);
+
+        // Ensure relevant relations are updated after delete.
+        utils::insert_emulated_on_delete(graph, query_schema, &model, &read_node, &delete_node)?;
```

Comment:
You want to be emulating referential actions even when atomic deletes can be done. Or else, you can avoid atomic deletes altogether as soon as the relationMode is not `foreignKeys` (but that would disable the feature entirely for MongoDB)

Reply:
It took some figuring out, but this should be working now. The only mode we cannot emulate with atomic deletes is `Restrict`, since it requires taking a look at the data before performing the operation (see comments in `query-engine/core/src/query_graph_builder/write/delete.rs`). So I implemented a check for `Restrict` and we have any (either in delete directly or in update by extension), we fall back to non-atomic deletes.